prereise/gather/griddata/hifld/data_process/transmission.py

```diff
@@ -27,6 +29,162 @@ def check_for_location_conflicts(substations):
         )
 
 
+def map_lines_to_substations_using_coords(substations, lines, **kwargs):
+    """Map lines to substations using coordinates
+
+    :param pandas.DataFrame substations: data frame of substations.
+    :param pandas.DataFrame lines: data frame of all lines.
+    :param int \*\*kwargs: optional arguments.
+    :return: (*pandas.DataFrame*) -- indices: lines id, columns: lines endpoint
+        coordinates (*FROM* and *TO*) and id of substations found at these coordinates (*FROM_SUB* and *TO_SUB*)
+    """
+    rounding = 3 if "rounding" not in kwargs else kwargs["rounding"]
+    subcoord2subid = (
+        substations.round(rounding).groupby(["LONGITUDE", "LATITUDE"]).groups
+    )
+    print(
+        f"initial number of substations: {substations.shape[0]}\n"
+        f"substations with same location after rounding: "
+        f"{substations.shape[0] - len(subcoord2subid)}\n"
+    )
+
+    lines_coord_from = lines["COORDINATES"].map(
+        lambda x: (round(x[0][0], rounding), round(x[0][1], rounding))
+    )
+    lines_coord_to = lines["COORDINATES"].map(
+        lambda x: (round(x[-1][0], rounding), round(x[-1][1], rounding))
+    )
+    line2sub = lines_coord_from.to_frame(name="FROM").join(
+        lines_coord_to.to_frame(name="TO")
+    )
+
+    for endpoint in ["FROM", "TO"]:
+        line2sub[f"{endpoint}_SUB"] = [
+            i if isinstance(i, float) else list(i)
```

Comment:
Is this ternary operator based on the logic that if a line endpoint isn't present in `subcoord2subid`, we'll get `NaN` which is a float, rather than an `int` index?

Reply:
After the mapping, we have in the `FROM_SUB` and `TO_SUB` some NA (`float` type) and some `pandas.Int64Index`. I cast the `Int64Index` to `list` since later I working with some `set`, `itertools`, etc.

Reply:
Do you think it would be clearer if we did something like:
```python
line2sub[f"{endpoint}_SUB"] = line2sub[f"{endpoint}"].apply(
    lambda x: list(subcoord2subid[x]) if x in subcoord2subid else pd.NA
)
```
?

Reply:
Also I guess we can simplify `f"{endpoint}"` to just `endpoint`.

Reply:
When I use your implementation I have got the following error later (l. 135)
`*{tuple(i) for j in ["FROM_SUB", "TO_SUB"] for i in line2sub[j]}`
TypeError: 'NAType' object is not iterable

Reply:
I think we need to change `pd.NA` into something iterable, like `(pd.NA,)` or `[pd.NA]` or `[]`.

Reply:
The funny thing is that in the previous implementation the NaN automatically generated by `map` were `float` and it was fine

Reply:
I think the difference is:
```
>>> import pandas as pd
>>> type(pd.NA)
<class 'pandas._libs.missing.NAType'>
```
and
```
>>> import numpy as np
>>> type(np.NAN)
<class 'float'>
```

Reply:
I don't know why the downstream behavior is different, since conceptually I think we are doing the same thing:
1. Loop through every entry in `line2sub[f"{endpoint}"]`.
2. If the entry is present in `subcoord2subid`, return the result wrapped in a `list`.
3. If not, return `float("nan")` or `pd.NA`.

Reply:
You can't `tuple` `float("nan")` either:
```python
>>> tuple(float("nan"))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'float' object is not iterable
```

Reply:
I think @rouille 's post tells the issue. @danielolsen Yeah, we can convert that into tuple, but we can construct a tuple with it: 
![image](https://user-images.githubusercontent.com/52716585/132765269-db288a04-c506-4092-99ff-37d789069102.png)

Reply:
This works:
```
line2sub[f"{endpoint}_SUB"] = line2sub[endpoint].apply(
            lambda x: list(subcoord2subid[x]) if x in subcoord2subid else np.NaN
```

Reply:
It's as simple as `np.NaN` vs `pd.NA`? That's spooky.

Reply:
I got the same issue with the `pandas.Int64Index` vs `list`

Reply:
Done